tools/cli/adminDBTimers.go

```diff
@@ -0,0 +1,169 @@
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+// SOFTWARE.
+
+package cli
+
+import (
+	"encoding/json"
+	"fmt"
+	"time"
+
+	"github.com/gocql/gocql"
+	"github.com/pkg/errors"
+	"github.com/urfave/cli"
+	"golang.org/x/sync/errgroup"
+
+	"github.com/uber/cadence/common"
+	"github.com/uber/cadence/common/backoff"
+	"github.com/uber/cadence/common/log/loggerimpl"
+	"github.com/uber/cadence/common/persistence"
+	cassp "github.com/uber/cadence/common/persistence/cassandra"
+	"github.com/uber/cadence/common/quotas"
+)
+
+// AdminTimers is used to list scheduled timers.
+func AdminTimers(c *cli.Context) {
+	lowerShardBound := c.Int(FlagLowerShardBound)
+	upperShardBound := c.Int(FlagUpperShardBound)
+	startDate := c.String(FlagStartDate)
+	endDate := c.String(FlagEndDate)
+	rps := c.Int(FlagRPS)
+
+	ctx, cancel := newContextForLongPoll(c)
+	defer cancel()
+
+	st, err := parseSingleTs(startDate)
+	if err != nil {
+		ErrorAndExit("wrong date format for "+FlagStartDate, err)
+	}
+	et, err := parseSingleTs(endDate)
+	if err != nil {
+		ErrorAndExit("wrong date format for "+FlagEndDate, err)
+	}
+
+	session := connectToCassandra(c)
+	defer session.Close()
+
+	// shared limiter
+	limiter := quotas.NewSimpleRateLimiter(rps)
+
+	group, ctx := errgroup.WithContext(ctx)
+	for shardID := lowerShardBound; shardID < upperShardBound; shardID++ {
+		shardID := shardID
+		group.Go(func() error {
+			return printTimers(c, session, shardID, st, et, limiter)
+		})
+	}
+
+	if err := group.Wait(); err != nil {
+		ErrorAndExit("getting timers", err)
+	}
+}
+
+func printTimers(
+	c *cli.Context,
+	session *gocql.Session,
+	shardID int,
+	startDate, endDate time.Time,
+	limiter *quotas.RateLimiter,
+) error {
+	domainID := c.String(FlagDomainID)
+	skipErrMode := c.Bool(FlagSkipErrorMode)
+	batchSize := c.Int(FlagBatchSize)
+	timerTypes := intSliceToMap(c.IntSlice(FlagTimerType))
+	logger := loggerimpl.NewNopLogger()
+
+	execStore, err := cassp.NewWorkflowExecutionPersistence(shardID, session, logger)
+	if err != nil {
+		err = errors.Wrapf(err, "cannot get execution manager for shardId %d", shardID)
+		if !skipErrMode {
+			return err
+		}
+		fmt.Println(err.Error())
+		return nil
+	}
+
+	manager := persistence.NewExecutionManagerImpl(execStore, logger)
+	ratelimitedClient := persistence.NewWorkflowExecutionPersistenceRateLimitedClient(manager, limiter, logger)
+
+	var token []byte
+	isFirstIteration := true
+	for isFirstIteration || len(token) != 0 {
+		isFirstIteration = false
+		req := persistence.GetTimerIndexTasksRequest{
+			MinTimestamp:  startDate,
+			MaxTimestamp:  endDate,
+			BatchSize:     batchSize,
+			NextPageToken: token,
+		}
+
+		resp := &persistence.GetTimerIndexTasksResponse{}
+
+		op := func() error {
+			var err error
+			resp, err = ratelimitedClient.GetTimerIndexTasks(&req)
+			return err
+		}
+
+		for attempt := 0; attempt < maxDBRetries; attempt++ {
```

Comment:
Can we rely on the retry policy itself (or may be create a new retry policy) here, instead of adding an extra layer of retry?